src/annotator.cpp

```diff
@@ -0,0 +1,361 @@
+            std::string prefix;
+            std::string reference;
+            double exponent;
+            double multiplier;
+            model->units(u)->unitAttributes(i, reference, prefix, exponent, multiplier, id);
+            if (!id.empty()) {
+                idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::UNIT, std::make_pair(model->units(u), i))));
+            }
+        }
+        if (model->units(u)->isImport() && model->units(u)->importSource() != nullptr) {
+            id = model->units(u)->importSource()->id();
+            idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::IMPORT, model->units(u)->importSource())));
+        }
+    }
+    // Components.
+    for (size_t c = 0; c < model->componentCount(); ++c) {
+        listComponentIdsAndItems(model->component(c), idList);
+    }
+    // Encapsulation.
+    id = model->encapsulationId();
+    if (!id.empty()) {
+        idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::ENCAPSULATION, std::string())));
+    }
+
+    return idList;
+}
+
+void Annotator::build(const ModelPtr &model)
+{
+    mPimpl->isBuilt = false;
+    removeAllIssues();
+    mPimpl->mIdList.clear();
+    mPimpl->mIdList = listIdsAndItems(model);
+    mPimpl->isBuilt = true;
+    mPimpl->mModel = model;
+}
+
+AnyItem Annotator::item(const std::string &id)
+{
+    AnyItem item;
+    if (!mPimpl->isBuilt) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Please call the Annotator::build function before attempting to access items by their id.");
+        issue->setLevel(libcellml::Issue::Level::ERROR);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+    if (mPimpl->mIdList.empty()) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("The id map is empty; the supplied model has no id attributes.");
+        issue->setLevel(libcellml::Issue::Level::ERROR);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+
+    // Retrieve the item from the idList.
+    ItemList::iterator it;
+    it = mPimpl->mIdList.find(id);
+
+    if (it == mPimpl->mIdList.end()) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Could not find an item with id='" + id + "' in the model.");
+        issue->setLevel(libcellml::Issue::Level::WARNING);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        // Adding the requested id string and this issue to the map, so that future requests (ie: to the .second)
+        // will return the same IssuePtr rather than creating a new one.
+        mPimpl->mIdList.insert(std::make_pair(id, item));
+        return item;
+    }
+    return mPimpl->mIdList[id];
+}
+
+ComponentPtr Annotator::component(const std::string &id)
```

Comment:
I can see these methods all look like they should work, but would be keen to know if the try/catch approach is what we should be using rather that testing the item type. @hsorby @agarny

Reply:
could be other reasons for the cast to fail beyond just the item type. So this seems like a reasonable way to go and will ensure that the casting exceptions would ever propagate out from the library